### ISSUE TITLE
Broken text alignment in SVG after Prettier formatting

### ISSUE THREAD

**Issue by lcnittl** [prettier/prettier#10121]
**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeA-ADwLYBsAEAbnAE4DOAltALwA6IAjAHQAM9+CkAJhVAOZ0QAVQAqAMQC0ADnZkYAQyhd5uaHEFQI9dAD5aUVGUJ99+MzlxQyggBYwYAByQB6ZwHcPTNwGYmEEnzOAEwsoc5GJiCmZjZwFHx2gkEAbNjY9NH4bhRcMDaC3gDsaRlQZvgWSGQO8mDq9A4kcGSkxKXlOYIRDN4sMlFlZsTkVFCCzAz0Oqgmg+UwJIpkAGb+2ILY8gsUmAAUzN6HR94ANCxnJxIHx4cXIQAsrAxSAJzeAJTt5fid9Hw9DGSU1Q8EwMEy3wAnoIrgwgkxkiwel9vhVxiwAKxMY6TAaosy-ECgmA9FIo75ySG4eogVawCSEeQkCiKGBITQkTa4ADcdJgEjccQSbIARhBcFxedB+ZQAF5wJAMDEOTBS+nLeTYCi4SFIADkAAURPgAMpLPXc4kSFTxKBIBCStzMmC8PgSbAQLgK3AkCQwEWWuBg61QMA2fz2pS87W4JAAYlCiZY0dwuAkEBqYAoMF1DBTaZIAFdqezoPKSBBuXIKwBrBWaKBwcnzRZWVacrpgFRwfaXBifEDTGDVRQQ1HQ+iw+GI5F4-EEriCYc1KCk+7N76YdFYnFTABK8Ts+BtfEbXFQzmXimml6DMBvfBvET0UBAJxAGZd0DIyFATIrbgGkyCA-igKhuPIkI-u+IqLGAdYwCamausgCyFnA75wNgIpwFwXpcAAMoofCFvIfBwGIaxbC6-DICA8iFjAEBviAdh4AA6jY2bNJmcAmiB2YUIQ2aQnRYBkNBIC8C0JAwAaix8JsyAargLTvgAVmQmAAEJwQhZrYHABG8HAykqGpICaZgJqutSACKhYQPAZmqRhIA1OQpB0WxuAsY0vAwOxOR5MgUjnO5FYtOxiwOHRjTNK0pnvgAjo58DyRmoH0WQEiNrhuEsU0qUUE08lkUpSAqRZLRaqhRZuZQ-D2WlpmVeZbkKCKQW5DYyBBO+CzyNqroAMIQGk8h0c0GIsYWLQiPIIqgVVbmEOhACSSgIIhYDMg4MAAIJKCaObUi5LQAL4XUAA)
```sh
--parser html
```

**Input:**
```html
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns="http://www.w3.org/2000/svg"
   height="26mm"
   width="37mm"
   xml:space="preserve"
   id="svg1308"
   version="1.1"><g
     transform="matrix(1.3333333,0,0,-1.3333333,0,204.01893)"
     id="g1316"><text
       y="-112.6013"
       x="105.33331"
       id="text1326"
       style="font-variant:normal;font-weight:bold;font-size:15px;font-family:'PT Sans';text-align:end;writing-mode:lr-tb;text-anchor:end;fill:#000000;fill-opacity:1;fill-rule:nonzero;stroke:none"
       transform="scale(1,-1)"><tspan
         y="-112.6013"
         id="tspan1324"
         x="105.33331">Right aligned</tspan></text></g></svg>

```

![image](https://user-images.githubusercontent.com/23134999/105612687-ccbbe480-5dbd-11eb-8b37-02c2be893aab.png)

**Output:**
```html
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
  xmlns="http://www.w3.org/2000/svg"
  height="26mm"
  width="37mm"
  xml:space="preserve"
  id="svg1308"
  version="1.1"
>
  <g transform="matrix(1.3333333,0,0,-1.3333333,0,204.01893)" id="g1316">
    <text
      y="-112.6013"
      x="105.33331"
      id="text1326"
      style="
        font-variant: normal;
        font-weight: bold;
        font-size: 15px;
        font-family: 'PT Sans';
        text-align: end;
        writing-mode: lr-tb;
        text-anchor: end;
        fill: #000000;
        fill-opacity: 1;
        fill-rule: nonzero;
        stroke: none;
      "
      transform="scale(1,-1)"
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">Right aligned</tspan>
    </text>
  </g>
</svg>

```

![image](https://user-images.githubusercontent.com/23134999/105612614-b01fac80-5dbd-11eb-9921-7270810a4ddc.png)

**Expected behavior:**

The rendered image should look the same before and after formatting. The screenshots above are from Inkscape, but the same behavior is seen when opening the files with Firefox.

I was not really able to track the reason for this down, however, I found that removal of `xml:space="preserve"` fixes the file:

```html
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
  xmlns="http://www.w3.org/2000/svg"
  height="26mm"
  width="37mm"
  id="svg1308"
  version="1.1"
>
  <g transform="matrix(1.3333333,0,0,-1.3333333,0,204.01893)" id="g1316">
    <text
      y="-112.6013"
      x="105.33331"
      id="text1326"
      style="
        font-variant: normal;
        font-weight: bold;
        font-size: 15px;
        font-family: 'PT Sans';
        text-align: end;
        writing-mode: lr-tb;
        text-anchor: end;
        fill: #000000;
        fill-opacity: 1;
        fill-rule: nonzero;
        stroke: none;
      "
      transform="scale(1,-1)"
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">Right aligned</tspan>
    </text>
  </g>
</svg>

```

![image](https://user-images.githubusercontent.com/23134999/105612783-64b9ce00-5dbe-11eb-86bc-45ea09e2642c.png)

So it has to do with some spacing I assume. Yet, the `style` key, whose value differs the most after formatting, is not the problem (I tried to use the value from the original, unforamtted file).

(This issue was also mentioned in #8907 - I have shortened the example from over there, which interestingly led to a less pronounced effect.)

**Comment by lcnittl** [comment]
I have now tracked down the reason, it is a newline between `</tspan>` and `</text>`. The following svg-file, with this newline removed, produces the correct output:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
  xmlns="http://www.w3.org/2000/svg"
  height="26mm"
  width="37mm"
  xml:space="preserve"
  id="svg1308"
  version="1.1"
>
  <g transform="matrix(1.3333333,0,0,-1.3333333,0,204.01893)" id="g1316">
    <text
      y="-112.6013"
      x="105.33331"
      id="text1326"
      style="
        font-variant: normal;
        font-weight: bold;
        font-size: 15px;
        font-family: 'PT Sans';
        text-align: end;
        writing-mode: lr-tb;
        text-anchor: end;
        fill: #000000;
        fill-opacity: 1;
        fill-rule: nonzero;
        stroke: none;
      "
      transform="scale(1,-1)"
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">Right aligned</tspan></text>
  </g>
</svg>
```

~~It is interesting, that the first newline within the `text` bock (between `<text ...>` and `<tspan ...>` does not cause problems.~~ → It does, but the spaces are invisible (cf. below)

Not an expert in SVG (and unfortunately no time to dig the specs): It appears to me that any character within a `text` block matters (newlines seem to be transformed to spaces). Some simple testing supports my assumption (snippets' first line is the closing bracket of `<text ...>`, the rest is taken from the svg-file from this comment):

Let's replace the body of `tspan` with `body` and put `tail` in the `text` block (as tail of `tspan`):

```xml
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">body</tspan>tail</text>
  </g>
```

![image](https://user-images.githubusercontent.com/23134999/105840006-9a370500-5fd2-11eb-8232-c4dd1d737df6.png)

The text is right aligned.

---

We now put a single line break between `ta` and `il`:

```xml
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">body</tspan>ta
il</text>
  </g>
```

![image](https://user-images.githubusercontent.com/23134999/105840067-b470e300-5fd2-11eb-87a0-3af61eefbcf2.png)

The text is still right aligned, there is space between `ta` and `il`.

---

We now indent the line that contains `B`:

```xml
    >
      <tspan y="-112.6013" id="tspan1324" x="105.33331">body</tspan>ta
    il</text>
  </g>
```

![image](https://user-images.githubusercontent.com/23134999/105840107-c5215900-5fd2-11eb-83a9-e2b730b5179e.png)

The text is still right aligned, there is 5 spaces between `ta` and `il`.

---

So in the end, the text from the original example is still right aligned, but there are 5 spaces (1 from `\n` + 4 from `space`) after the text making it appear not right aligned anymore.

As this also applies to the spaces between `<text ...>` and `<tspan ...>` this line should also be not broken. The desired outcome would thus be:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
  xmlns="http://www.w3.org/2000/svg"
  height="26mm"
  width="37mm"
  xml:space="preserve"
  id="svg1308"
  version="1.1"
>
  <g transform="matrix(1.3333333,0,0,-1.3333333,0,204.01893)" id="g1316">
    <text
      y="-112.6013"
      x="105.33331"
      id="text1326"
      style="
        font-variant: normal;
        font-weight: bold;
        font-size: 15px;
        font-family: 'PT Sans';
        text-align: end;
        writing-mode: lr-tb;
        text-anchor: end;
        fill: #000000;
        fill-opacity: 1;
        fill-rule: nonzero;
        stroke: none;
      "
      transform="scale(1,-1)"
    ><tspan y="-112.6013" id="tspan1324" x="105.33331">Right aligned</tspan></text>
  </g>
</svg>
```

Not sure, why Inkscape even puts a `tspan` block there, as it works without as well – but yes, they certainly have their reasons.

**Comment by thorn0** [comment]
Thanks for digging into this. Looks like the solution would be to switch to the `--html-whitespace-sensitivity strict` mode for SVG elements that have the `xml:space="preserve"` attributes and for their descendants.

**Comment by fisker** [comment]
I think it's more like `pre` https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/xml:space

**Comment by thorn0** [comment]
Yes, it is, but what I wrote still will be better than how it works now. Unless #7393 gets fixed.

**Comment by brodybits** [comment]
I would favor getting #7393 fixed, if possible.

**Comment by thorn0** [comment]
@brodybits Who wouldn't? But the `strict` whitespace mode is already implemented and using it is a good enough and (hopefully) quick way to make this problem occur significantly less often. Whereas implementing a "`pre` mode" properly is quite an endeavor.

**Comment by vincerubinetti** [comment]
I think this is related, but I have an SVG like this:

```
<text>
  <tspan>Some</tspan>
  <tspan>Text</tspan>
</text>
```

And I'm looking for the tspans to be "continuous", but Prettier puts them onto separate lines and creates an extra whitespace. It's weird because when working with JSX or something, I feel like Prettier usually preserves the HTML whitespace quirks with something like this (but nicer looking):

```
<text>
  <tspan>Some</tspan><tspan
   >Text</tspan>
</text>
```

But it doesn't do that here for some reason.

**Comment by mikestopcontinues** [comment]
Yes, `<tspan>` should really stay on the same line, especially in 'strict', but it doesn't.
